### PR TITLE
feat: `/api/legacy/*` wrappers for three ES-compatible RPC endpoints

### PR DESCRIPTION
**Pull request by akolotov**
## What

Adds three thin GET routes under `/api/legacy/*` that republish existing Etherscan-compatible (ES-compatible) RPC endpoints as routes with full OpenAPI specifications.

| New route | Bridges to |
|---|---|
| `GET /api/legacy/logs/get-logs` | `GET /api?module=logs&action=getlogs` |
| `GET /api/legacy/block/get-block-number-by-time` | `GET /api?module=block&action=getblocknobytime` |
| `GET /api/legacy/block/eth-block-number` | `GET /api?module=block&action=eth_block_number` |

The existing `/api?module=…&action=…` endpoints are **not changed, not deprecated, and not removed**.

## Why

Blockscout has a formal OpenAPI document at `/api/v2/openapi`. The Etherscan-compatible RPC endpoints at `/api?module=…` have **no OpenAPI representation** — they are effectively invisible to any tool that discovers APIs by reading a spec, including AI agents.

An agent querying Blockscout today can see the full spec surface and pick endpoints from it confidently. Anything only available via the query-string dispatch mechanism is invisible; the agent will either fail a query it could have answered or synthesise it from the wrong endpoint. The three endpoints in this PR are exactly the "specific data, one endpoint" primitives agents need most: filtered event logs, block number by timestamp, and latest block number.

This is an **observability change**, not a migration. No existing client needs to move.

## How it works

Each new controller action is a one-liner that delegates to the corresponding v1 controller function. The response body is JSON-equal to the v1 endpoint; only the route and the OpenAPI registration differ.

Key design decisions:

- **No `CastAndValidate` plug.** All query params are declared `optional` in the spec. Required-param enforcement and error responses stay in the v1 controllers, preserving the `{"status":"0","message":"…","result":null}` envelope shape for invalid input instead of replacing it with a 422.
- **Delegate view modules** (`defdelegate render/2`) wire Phoenix's automatic view lookup to the v1 view without needing `put_view/2` in every action. `BlockView` adds one explicit clause to bridge the action-derived template name (`get_block_number_by_time.json`) to the v1 template name (`getblocknobytime.json`), since the v1 controller calls `render/2` without an explicit template on its success path.
- **Inline routes**, not a sub-router. `Paths.from_router/1` in `specs/public.ex` auto-detects them; no manual registration needed.
- **Envelope schemas as helper functions** (`Envelope.rpc_envelope/1`, `Envelope.eth_rpc_envelope/1`) following the `paginated_response/1` precedent. Acceptable for three endpoints; promote to `OpenApiSpex.schema` modules before the legacy surface grows past ~5 endpoints.

## New files

| File | Purpose |
|---|---|
| `controllers/api/legacy/logs_controller.ex` | `get_logs` action + OpenAPI operation |
| `controllers/api/legacy/block_controller.ex` | `get_block_number_by_time` + `eth_block_number` actions + OpenAPI operations |
| `views/api/legacy/logs_view.ex` | Delegates all templates to `API.RPC.LogsView` |
| `views/api/legacy/block_view.ex` | Bridges `get_block_number_by_time.json` → `getblocknobytime.json`; delegates the rest to `API.RPC.BlockView` |
| `schemas/api/legacy/envelope.ex` | `rpc_envelope/1` and `eth_rpc_envelope/1` schema helpers |
| `schemas/api/legacy/log_item.ex` | Full `OpenApiSpex.schema` for a single log entry |
| `schemas/api/legacy/get_block_number_by_time_result.ex` | Schema for `{blockNumber: decimal-string}` |
| `schemas/api/legacy/eth_block_number_result.ex` | Schema for hex-encoded block number string |

## Tests

- **Controller tests** cover all error paths (missing params, invalid format, not-found), success paths, edge cases (`fromBlock=latest`, two-topic-without-opr), and **parity assertions** — each test issues the same request to both the v1 and the legacy endpoint and asserts `json_response(v1, 200) == json_response(legacy, 200)`.
- **`public_legacy_tag_test.exs`** — spec-presence invariants: for each of the three paths, asserts the path exists in `Specs.Public.spec()`, carries `tags: ["legacy"]`, has a declared 200 response, and that 200 response has an `application/json` schema. Guards against the silent failure mode where a route appears in the spec but has no documented response.
- Every `json_response/2` call in the test suite automatically validates the response body against the OpenAPI schema via `TestApiSchemaAssertions`.

## OpenAPI changes

Three new GET operations appear under the `legacy` tag in `/api/v2/openapi`. No existing operations are modified.

```
GET /legacy/logs/get-logs
GET /legacy/block/get-block-number-by-time
GET /legacy/block/eth-block-number
```

## Checklist for your Pull Request (PR)

- [x] I verified this PR does not break any public APIs, contracts, or interfaces that external consumers depend on.
- [x] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I updated documentation if needed:
  - [ ] General docs: submitted PR to [docs repository](https://github.com/blockscout/docs).
  - [ ] ENV vars: updated [env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables) and set version parameter to `master`.
  - [ ] Deprecated vars: added to [deprecated env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables/deprecated-env-variables).
- [x] If I modified API endpoints, I updated the Swagger/OpenAPI schemas accordingly and checked that schemas are asserted in tests, and highlighted the change in the PR description.
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Summary by CodeRabbit

* **New Features**
  * Added three legacy API endpoints with backward-compatible routing and view adapters for logs, get-block-number-by-time, and eth-block-number.
* **Documentation**
  * Added legacy response envelope and result schemas; updated OpenAPI spec to include a "legacy" tag and normalized operation tag names.
* **Tests**
  * Added comprehensive parity and validation tests covering legacy endpoints and OpenAPI path presence.

### CODE REVIEW

**Review by nikitosing — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

File: apps/block_scout_web/lib/block_scout_web/routers/api_router.ex [hunk 1]
```diff
@@ -484,6 +484,19 @@ defmodule BlockScoutWeb.Routers.ApiRouter do
...
+
+    scope "/logs" do
+      get("/get-logs", Legacy.LogsController, :get_logs)
+    end
+
+    scope "/block" do
+      get("/get-block-number-by-time", Legacy.BlockController, :get_block_number_by_time)
+      get("/eth-block-number", Legacy.BlockController, :eth_block_number)
+    end
+  end
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/routers/api_router.ex:498]: Intentional — the scope is deliberately at `/api/legacy` rather than `/api/v2/legacy`. These routes wrap the _Etherscan-compatible_ (query-string) API, which predates and sits outside the v2 surface. Placing them under `/v2` would imply they are v2 endpoints, which they are not. The `/api/v2/openapi` spec documents all routes from `ApiRouter` regardless of prefix, so the legacy operations are still fully discoverable in the spec.

File: apps/block_scout_web/test/block_scout_web/controllers/api/legacy/block_controller_test.exs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        |> get("/api", Map.merge(params, %{"module" => "block", "action" => "getblocknobytime"}))
+        |> json_response(200)
+
+      v2_response =
+        conn
+        |> get("/api/legacy/block/get-block-number-by-time", params)
+        |> json_response(200)
+
+      assert v1_response == v2_response
+    end
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/test/block_scout_web/controllers/api/legacy/block_controller_test.exs:140]: JSON equality is intentional — the goal is to verify the delegating controller returns the same decoded data as the v1 endpoint, which `json_response/2` covers exactly. Byte-level parity is not a requirement here (key ordering or whitespace differences between the two handlers would be a false failure). The PR description has been updated to say 'JSON-equal' instead of 'byte-identical' to reflect this.

File: apps/block_scout_web/test/block_scout_web/controllers/api/legacy/logs_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        |> get("/api", Map.merge(params, %{"module" => "logs", "action" => "getLogs"}))
+        |> json_response(200)
+
+      v2_response =
+        conn
+        |> get("/api/legacy/logs/get-logs", params)
+        |> json_response(200)
+
+      assert v1_response == v2_response
+    end
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/test/block_scout_web/controllers/api/legacy/logs_controller_test.exs:244]: JSON equality is intentional — same reasoning as the sibling comment on `block_controller_test.exs`: `json_response/2` verifies the decoded payload matches, which is the actual invariant we care about. The PR description has been updated to say 'JSON-equal' instead of 'byte-identical'.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/legacy/envelope.ex [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        message: %Schema{
+          type: :string,
+          description:
+            "Human-readable status string — `OK` on success, " <>
+              "a descriptive error message otherwise."
+        },
+        result: %Schema{
+          description: "Endpoint-specific payload on success; `null` on error.",
+          allOf: [result_schema]
+        }
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/legacy/envelope.ex:30]: Fixed — applied the suggested patch. The `result` field now uses `anyOf: [%Schema{type: :null}, %Schema{allOf: [result_schema]}]` so the schema correctly models both the success payload and the `null` returned on error. Tests (12/12), code quality, and Dialyzer all pass.

[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/legacy/envelope.ex:30]: Correction to my previous reply: used `nullable: true` (OpenAPI 3.0 convention) rather than `anyOf: [%Schema{type: :null}, ...]` — that pattern is OpenAPI 3.1 syntax and this spec is 3.0. The result field is now `nullable: true, allOf: [result_schema]`, consistent with how every other nullable field in the codebase is modelled. Tests (12/12), code quality, and Dialyzer all pass.

File: apps/block_scout_web/lib/block_scout_web/specs/public.ex [hunk 5]
```diff
@@ -5,11 +5,62 @@ defmodule BlockScoutWeb.Specs.Public do
...
+    "token-transfers",
+    "smart-contracts",
+    "config",
+    "main-page",
+    "search",
+    "stats",
+    "csv-export",
+    "account-abstraction",
+    "withdrawals"
+  ]
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/specs/public.ex:34]: Why `default_api_categories` are strings, but `chain_type_category` and legacy category are `%Tag{}` struct

Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/specs/public.ex:34]: I renamed `chain_type_category` to `chain_type_category_tags` to eliminate confusion in https://github.com/blockscout/blockscout/pull/14239/changes/cfe14d75a98a6a93f5b41194b51cb8467043126d.

`default_api_categories` are converted to tags via:
```
Enum.map(@default_api_categories, fn category -> %Tag{name: category} end)
```

Comment (nikitosing) [reply apps/block_scout_web/lib/block_scout_web/specs/public.ex:34]: Ah, I see now, didn't notice that. Thanks

File: apps/block_scout_web/lib/block_scout_web/specs/public.ex [hunk 6]
```diff
@@ -5,11 +5,62 @@ defmodule BlockScoutWeb.Specs.Public do
...
+      end
+
+    {chain_type, nil} when chain_type in [:scroll, :zilliqa] ->
+      @chain_type_category_tags [%Tag{name: to_string(chain_type)}]
+      defp chain_type_category_tags, do: @chain_type_category_tags
+
+    _ ->
+      @chain_type_category_tags []
+      defp chain_type_category_tags, do: @chain_type_category_tags
+  end
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/specs/public.ex:62]: No, in fact, `mud` is a set of features for a redstone chain based on the Optimism rollup, while `celo` is a fork of Optimism with its own set of features. The feature sets of mud and celo do not overlap.
